### PR TITLE
[ARCTIC][AMS]: Add the metric statistics of a table

### LINKED ISSUE #477: [Feature][AMS]: Add the metric statistics of a table
### Description

Add table metric statistics, such as file size and file quantity

### Use case/motivation

Add table metric statistics, such as file size and file quantity

### Describe the solution

The metric of the table is collected periodically and written to the database

### Related issues

subtask of https://github.com/NetEase/arctic/issues/405

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hameizi**
## Why are the changes needed?

resolve #477

## Brief change log

add and done logic:
1.report table metric by tracer
2.report optimizer metric by optimizer
need add logic:
3.Periodically collect metric from the native iceberg table

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (no)
  - If yes, how is the feature documented? ( not documented)

### CODE REVIEW

File: optimizer/src/main/java/com/netease/arctic/optimizer/metric/TaskRecorder.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+   *
+   * @param outputFiles - outputFiles of this task
+   */
+  public void finishCurrentTask(Iterable<TaskStat.FileStat> outputFiles) {
+    finishCurrentTask(outputFiles, System.currentTimeMillis());
+  }
+
+  synchronized void finishCurrentTask(Iterable<TaskStat.FileStat> outputFiles, long endTime) {
+    if (this.currentTaskStat == null) {
+      LOG.warn("current task is null, ignore finish");
```
Comment (Aireed) [inline optimizer/src/main/java/com/netease/arctic/optimizer/metric/TaskRecorder.java:77]: we shall return here for this.currentTaskStat is null

File: optimizer/src/main/java/com/netease/arctic/optimizer/metric/TaskRecorder.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+  private double calculateUsage(long begin, long end, List<TaskStat> taskStats) {
+    if (end < begin) {
+      LOG.warn("end {} is not after begin {}, return usage 0.0", end, begin);
+      return 0.0;
+    }
+    if (end == begin) {
+      return 0.0;
+    }
+    if (taskStats.isEmpty()) {
```
Comment (Aireed) [inline optimizer/src/main/java/com/netease/arctic/optimizer/metric/TaskRecorder.java:124]: should we check whether taskStats is null?

**Review by Aireed — CHANGES_REQUESTED** [review]

**Comment by Aireed** [comment]
CC @hzluting  @wangtaohz  to review optimize part!

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  public void insertStatistic(TableMetricsStatistic fileSizeStatistic) {
+    try (SqlSession sqlSession = getSqlSession(true)) {
+      TableMetricsStatisticMapper metricsStatisticMapper = getMapper(sqlSession, TableMetricsStatisticMapper.class);
+      if (metricsStatisticMapper.getMetricsStatistic(fileSizeStatistic).size() > 0) {
+        metricsStatisticMapper.updateMetricsStatistic(fileSizeStatistic);
+      } else {
+        metricsStatisticMapper.insertMetricsStatistic(fileSizeStatistic);
+      }
+    } catch (Exception e) {
+      System.out.println("error when insert");
```
[the remark this answers is not in the document]
Comment (hameizi) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java:181]: will delete in new commit

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+      LOG.error("", e);
```
[the remark this answers is not in the document]
Comment (hameizi) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java:182]: will add in new commit

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    }
+  }
+
+  public void summaryMetrics() {
+    long statisticTime = System.currentTimeMillis() / 60000 * 60000;
```
[the remark this answers is not in the document]
Comment (hameizi) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java:187]: in order to get integral point for every minutes.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+  public static class TableMetricsStatisticTask implements Runnable {
+
+    private final MetricsStatisticService metricsStatisticService;
+
+    public TableMetricsStatisticTask() {
+      this.metricsStatisticService = ServiceContainer.getMetricsStatisticService();
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(TableMetricsStatisticTask.class);
```
[the remark this answers is not in the document]
Comment (hameizi) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/MetricsStatisticService.java:251]: will move in new commit.

File: ams/ams-server/src/test/java/com/netease/arctic/ams/server/AmsTestBase.java [hunk 7]
```diff
@@ -97,28 +100,29 @@
...
+    // TestIcebergMinorOptimizePlan.class,
+    // TestIcebergMajorOptimizeCommit.class,
+    // TestIcebergMinorOptimizeCommit.class,
+    // TestExpireFileCleanSupportIceberg.class,
+    // TestOrphanFileCleanSupportIceberg.class,
+    // TestOrphanFileClean.class,
+    // TestFileInfoCacheService.class,
+    // SupportHiveTestGroup.class,
+    // TestArcticTransactionService.class,
+    // TestOptimizerService.class,
```
[the remark this answers is not in the document]
Comment (hameizi) [reply ams/ams-server/src/test/java/com/netease/arctic/ams/server/AmsTestBase.java:124]: will fix in new commit.

**Pull request closed without merging**